alacritty_terminal/src/term/cell.rs

```diff
@@ -47,7 +50,244 @@ impl<T: Copy> ResetDiscriminant<T> for T {
+        Self { mask: ColorMask::new(), fg, bg, underline }
+    }
+}
+
+impl PartialEq for CellColors {
+    fn eq(&self, other: &Self) -> bool {
+        self.mask == other.mask
+            && self.fg() == other.fg()
+            && self.bg() == other.bg()
+            && self.underline() == other.underline()
+    }
+}
+impl Eq for CellColors {}
+
+impl Serialize for CellColors {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut s = serializer.serialize_struct("CellColors", 2)?;
+        s.serialize_field("fg", &self.fg())?;
+        s.serialize_field("bg", &self.bg())?;
+        s.serialize_field("underline", &self.underline())?;
+        s.end()
+    }
+}
+
+impl<'de> Deserialize<'de> for CellColors {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        struct UnpackedColors {
+            fg: Color,
+            bg: Color,
+            underline: Option<Color>,
+        }
+
+        let unpacked = UnpackedColors::deserialize(deserializer)?;
+        let mut cell_colors = CellColors::default();
+        cell_colors.set_fg(unpacked.fg);
+        cell_colors.set_bg(unpacked.bg);
+        cell_colors.set_underline_color(unpacked.underline);
+        Ok(cell_colors)
+    }
+}
+
+impl Color {
+    #[inline]
+    fn into_packed_with_type(self) -> (ColorTy, PackedColor) {
+        match self {
+            Self::Spec(rgb) => (ColorTy::SPEC, PackedColor { rgb }),
+            Self::Indexed(indexed) => (ColorTy::INDEXED, PackedColor { indexed }),
+            Self::Named(named) => (ColorTy::NAMED, PackedColor { named }),
+        }
+    }
+}
+
+impl CellColors {
+    #[inline]
+    pub fn set_fg(&mut self, color: Color) {
+        let (ty, color) = color.into_packed_with_type();
+        self.fg = color;
+        self.mask.set_fg_ty(ty);
+    }
+
+    #[inline]
+    pub fn set_bg(&mut self, color: Color) {
+        let (ty, color) = color.into_packed_with_type();
+        self.bg = color;
+        self.mask.set_bg_ty(ty);
+    }
+
+    #[inline]
+    pub fn set_underline_color(&mut self, color: Option<Color>) {
+        let (ty, color) = match color {
+            Some(color) => color.into_packed_with_type(),
+            None => {
+                self.mask.clear_underline();
+                return;
+            },
+        };
+        self.underline = color;
+        self.mask.set_underline_ty(ty);
+    }
+
+    #[inline]
+    pub fn fg(&self) -> Color {
+        self.fg.into_color(self.mask.fg_ty())
+    }
+
+    #[inline]
+    pub fn bg(&self) -> Color {
+        self.bg.into_color(self.mask.bg_ty())
+    }
+
+    #[inline]
+    pub fn underline(&self) -> Option<Color> {
+        if self.mask.has_underline() {
+            Some(self.underline.into_color(self.mask.underline_ty()))
+        } else {
+            None
+        }
```

Comment:
Why bother returning an Option when the underline will always be the fg otherwise anyway? No reason to do this at the callee site, right?